### DISCUSSION TITLE
Unable to scale shapes in a canvas / How to position shapes outside of a Canvas

### DISCUSSION THREAD

**Discussion by Mrcubix** [AvaloniaUI/Avalonia#13317] [Q&A]
Currently working on a responsive app, one of the control right now is a node canvas, where nodes of different shapes can be placed, moved, and are sized using respectively the user's cursor & a couple of input boxes.

Unfortunately, it does not seem like the canvas scale its elements by default at least.

It does not seem like it has a stretch property either, which could allow it to act like an image for example, stretching its content.
Instead the result is a bunch static shapes (that does not move not scale with window size).

Grid however, can change the visual location of its elements when changing the windows size, however, the size of the elements stays static, which is still problematic.

Would there be a way to achieve such a thing without having to rely on custom drawn controls?

The code in question:
```cs
          <ItemsControl ItemsSource="{TemplateBinding Nodes}">

            <ItemsControl.ItemContainerTheme>
              <ControlTheme TargetType="ContentPresenter" x:DataType="nodes:NodeViewModel">
                <Setter Property="Canvas.Left" Value="{Binding X}"/>
                <Setter Property="Canvas.Top" Value="{Binding Y}"/>
              </ControlTheme>
            </ItemsControl.ItemContainerTheme>

            <ItemsControl.ItemsPanel>
              <ItemsPanelTemplate>
                <Canvas/>
              </ItemsPanelTemplate>
            </ItemsControl.ItemsPanel>

            <ItemsControl.DataTemplates>
              <DataTemplate DataType="nodes:RectangleNodeViewModel">
                <gn:RectangleNode Fill="Blue" Stroke="Black" StrokeThickness="2"
                                  Width="{Binding Width}" Height="{Binding Height}" />
              </DataTemplate>
              <DataTemplate DataType="nodes:CircleNodeViewModel">
                <gn:CircleNode Fill="Blue" Stroke="Black" StrokeThickness="2"
                               Width="{Binding Width}" Height="{Binding Height}" />
              </DataTemplate>
            </ItemsControl.DataTemplates>

          </ItemsControl>
```
The parent for this TemplatedControl is a Grid, the column in which this control is wrapped in is defined as follow :
```axaml
<ColumnDefinition Width="*" />
```

**Comment by stevemonaco** [comment]
Wrap the `Canvas` in a `Viewbox`. That will stretch the `Canvas` to match the parent dimensions. For zooming/panning over elements, you might apply a `ScaleTransform` and `TranslateTransform` to the `Canvas`.

You might check out [Node2D](https://github.com/wieslawsoltes/NodeEditor) for some direction or use [PanAndZoom](https://github.com/wieslawsoltes/PanAndZoom) if you're building mostly from scratch.

**Reply by Mrcubix** [reply]
I'm about sure that the usage of viewbox is a bandaid solution to my problem so i want to avoid that

**Reply by timunie** [reply]
LayoutTransformControl

**Reply by Mrcubix** [reply]
> LayoutTransformControl

Forgot to mention that the canvas is used as a panel for ItemsControl, there this won't work.

**Reply by timunie** [reply]
Why? You could use that inside a ControlTemplate for ItemsControl from my understanding.

**Reply by Mrcubix** [reply]
> Why ?

Like an error occurring telling you ItemsControl only accept panels ?

> You could use that inside a ControlTemplate for ItemsControl from my understanding.

I'll need you to give me an example to understand what you're talking about, since the documentation for this element is basically inexistant.
And proper documentation from WPF and UWP does not exist for this element either.

- Just wrapping the `ItemsControl` inside a `LayoutTransformControl` with an empty `ScaleTransform` in `LayoutTransformControl.LayoutTransform` does nothing.
- `ItemControls.ItemPanel` requires a panel, which `LayoutTransformControl` isn't

**Reply by timunie** [reply]
https://github.com/AvaloniaUI/Avalonia/blob/fcea3453ff62225c4be56cf2569edffc72e7402e/src/Avalonia.Themes.Fluent/Controls/ItemsControl.xaml#L5-L15

**Comment by robloo** [comment]
You need a control specifically for this case that also supports zoom (scaling). That isn't provided out of the box but there are some controls in existence as @stevemonaco stated: https://github.com/wieslawsoltes/PanAndZoom

I usually see this type of thing done with custom drawn controls though. It's more complicated to setup but also gives way more control for advanced editing.

**Reply by Mrcubix** [reply]
I'm well aware that i could just make a custom drawn control based on the Image control 
(I have done that for my Color Tiles game when i couldn't get Uniform Grid to work the way i wanted)
but i would want to avoid that if more maintainable means exist.

\+ panning and zooming isn't what i'm looking for, what i'm looking for is stretching, like an image.

**Comment by gaino1122** [comment]
`<Viewbox Grid.Row= "1">
        <Grid Grid.Row="1">
            <Image Source="avares://" Stretch="Fill"/>
            <Canvas x:Name="CanvasArea" Background="Transparent">
                
            </Canvas>
        </Grid>
</Viewbox>`

Worked and fixed my resize issue Thanks
